test/unit/scan/test_scan_policies.py

```diff
@@ -7,14 +7,207 @@
+            fix_windows_path_limit(str(d / "symlink_dir" / "symlink_dir" / "file1.txt")),
+            fix_windows_path_limit(
+                str(d / "symlink_dir" / "symlink_dir" / "symlink_dir" / "file1.txt")
+            ),
+        ]
+
+    @pytest.mark.skipif(
+        platform.system() == 'Windows' and platform.python_implementation() == 'PyPy',
+        reason="Symlinks not supported on PyPy/Windows"
+    )
+    @pytest.mark.timeout(5)  # Set a 5-second timeout for this test
+    def test_circular_symlink_with_limit_1(self, tmp_path):
+        d = tmp_path / "dir"
+        d.mkdir()
+
+        (d / "file1.txt").write_text("content1")
+
+        symlink_dir = d / "symlink_dir"
+        symlink_dir.symlink_to(d)
+
+        policies_manager = ScanPoliciesManager()
+        policies_manager.max_symlink_visits = 1
+
+        folder = LocalFolder(str(tmp_path))
+        local_paths = folder.all_files(reporter=MagicMock(), policies_manager=policies_manager)
+        absolute_paths = [path.absolute_path for path in list(local_paths)]
+
+        assert len(absolute_paths) == 2
+        assert absolute_paths == [
+            fix_windows_path_limit(str(d / "file1.txt")),
+            fix_windows_path_limit(str(d / "symlink_dir" / "file1.txt")),
+        ]
+
+    # Just for fun, another test, based on a comment by @zackse available here:
+    # https://github.com/Backblaze/B2_Command_Line_Tool/issues/513#issuecomment-426751216
+
+    @pytest.mark.skipif(
+        platform.system() == 'Windows' and platform.python_implementation() == 'PyPy',
+        reason="Symlinks not supported on PyPy/Windows"
+    )
+    @pytest.mark.timeout(5)  # Set a 5-second timeout for this test
+    def test_circular_interesting_case(self, tmp_path):
+
+        # Create directories
+        o = tmp_path / "outsidedir/four/five/six"
+        o.mkdir(parents=True)
+
+        s = tmp_path / "startdir/one/two/three"
+        s.mkdir(parents=True)
+
+        # Create symbolic links
+        symlink_dir = tmp_path / "outsidedir/four/five/one"
+        symlink_dir.symlink_to(tmp_path / "startdir/one")
+
+        symlink_dir = tmp_path / "startdir/badlink"
+        symlink_dir.symlink_to("/")
+
+        symlink_dir = tmp_path / "startdir/outsidedir"
+        symlink_dir.symlink_to(tmp_path / "outsidedir")
+
+        # Create text files
+        file = tmp_path / "outsidedir/hello.txt"
+        file.write_text("hello")
+
+        file = tmp_path / "startdir/hello.txt"
+        file.write_text("hello")
+
+        file = tmp_path / "startdir/one/goodbye.txt"
+        file.write_text("goodbye")
+
+        folder = LocalFolder(str(tmp_path))
+
+        policies_manager = ScanPoliciesManager()
+        policies_manager.max_symlink_visits = 3
+
+        local_paths = folder.all_files(reporter=MagicMock(), policies_manager=policies_manager)
+        absolute_paths = [path.absolute_path for path in list(local_paths)]
+
+        assert len(absolute_paths) == 6
```

Comment:
```suggestion
```
the assert below already checks list length. Moreover if this assert fails, you'll get something like `3 != 2` which is way less useful than the feedback from the assert below.